sprint 22 playlist

pyclan88 opened the pull request:
большой 22 спринт

pyclan88 committed:
sprint 22 playlist

DmitrijEgorow commented:
- [x] Приложение аварийно завершается

https://github.com/user-attachments/assets/d03b3561-482f-460f-a45f-d4fa59daa0dc

```
FATAL EXCEPTION: main  
Process: com.practicum.playlistmaker, PID: 16800  
java.lang.NullPointerException  
at com.practicum.playlistmaker.medialibrary.ui.newplaylist.NewPlaylistFragment.getBinding(NewPlaylistFragment.kt:30)  
at com.practicum.playlistmaker.medialibrary.ui.newplaylist.NewPlaylistFragment.hasUnsavedChanges(NewPlaylistFragment.kt:148)  
at com.practicum.playlistmaker.medialibrary.ui.newplaylist.NewPlaylistFragment.handleBackNavigation(NewPlaylistFragment.kt:140)  
at com.practicum.playlistmaker.medialibrary.ui.newplaylist.NewPlaylistFragment.access$handleBackNavigation(NewPlaylistFragment.kt:26)  
at com.practicum.playlistmaker.medialibrary.ui.newplaylist.NewPlaylistFragment$handleOnBackPressed$1.handleOnBackPressed(NewPlaylistFragment.kt:99)  
at androidx.activity.OnBackPressedDispatcher.onBackPressed(OnBackPressedDispatcher.kt:279)  
at androidx.activity.ComponentActivity.onBackPressed(ComponentActivity.kt:612)  
at android.app.Activity.onKeyUp(Activity.java:3995)  
at android.view.KeyEvent.dispatch(KeyEvent.java:3532)  
at android.app.Activity.dispatchKeyEvent(Activity.java:4347)  
at androidx.core.app.ComponentActivity.superDispatchKeyEvent(ComponentActivity.kt:103)  
at androidx.core.view.KeyEventDispatcher.dispatchKeyEvent(KeyEventDispatcher.java:85)  
at androidx.core.app.ComponentActivity.dispatchKeyEvent(ComponentActivity.kt:117)  
at androidx.appcompat.app.AppCompatActivity.dispatchKeyEvent(AppCompatActivity.java:604)  
at androidx.appcompat.view.WindowCallbackWrapper.dispatchKeyEvent(WindowCallbackWrapper.java:59)  
at androidx.appcompat.app.AppCompatDelegateImpl$AppCompatWindowCallback.dispatchKeyEvent(AppCompatDelegateImpl.java:3397)  
at com.android.internal.policy.DecorView.dispatchKeyEvent(DecorView.java:928)  
at android.view.ViewRootImpl$ViewPostImeInputStage.processKeyEvent(ViewRootImpl.java:8056)  
at android.view.ViewRootImpl$ViewPostImeInputStage.onProcess(ViewRootImpl.java:7864)  
at android.view.ViewRootImpl$InputStage.deliver(ViewRootImpl.java:7213)  
at android.view.ViewRootImpl$InputStage.onDeliverToNext(ViewRootImpl.java:7270)  
at android.view.ViewRootImpl$InputStage.forward(ViewRootImpl.java:7236)  
at android.view.ViewRootImpl$AsyncInputStage.forward(ViewRootImpl.java:7434)  
at android.view.ViewRootImpl$InputStage.apply(ViewRootImpl.java:7244)  
at android.view.ViewRootImpl$AsyncInputStage.apply(ViewRootImpl.java:7491)  
at android.view.ViewRootImpl$InputStage.deliver(ViewRootImpl.java:7217)  
at android.view.ViewRootImpl$InputStage.onDeliverToNext(ViewRootImpl.java:7270)  
at android.view.ViewRootImpl$InputStage.forward(ViewRootImpl.java:7236)  
at android.view.ViewRootImpl$InputStage.apply(ViewRootImpl.java:7244)  
at android.view.ViewRootImpl$InputStage.deliver(ViewRootImpl.java:7217)  
at android.view.ViewRootImpl$InputStage.onDeliverToNext(ViewRootImpl.java:7270)  
at android.view.ViewRootImpl$InputStage.forward(ViewRootImpl.java:7236)  
at android.view.ViewRootImpl$AsyncInputStage.forward(ViewRootImpl.java:7467)  
at android.view.ViewRootImpl$ImeInputStage.onFinishedInputEvent(ViewRootImpl.java:7676)  
at android.view.inputmethod.InputMethodManager$PendingEvent.run(InputMethodManager.java:3897)  
at android.view.inputmethod.InputMethodManager.invokeFinishedInputEventCallback(InputMethodManager.java:3414)  
at android.view.inputmethod.InputMethodManager.finishedInputEvent(InputMethodManager.java:3405)  
at android.view.inputmethod.InputMethodManager$ImeInputEventSender.onInputEventFinished(InputMethodManager.java:3874)  
at android.view.InputEventSender.dispatchInputEventFinished(InputEventSender.java:154)  
at android.os.MessageQueue.nativePollOnce(Native Method)  
at android.os.MessageQueue.next(MessageQueue.java:335)  
at android.os.Looper.loopOnce(Looper.java:186)  
at android.os.Looper.loop(Looper.java:313)
```

DmitrijEgorow commented:
- [x] Клавиатура подпрыгивает и резко исчезает – странное поведение

https://github.com/user-attachments/assets/9f75c6db-72a9-4731-9ef6-bf2a6470cd07

DmitrijEgorow commented:
- [x] "Добавлено в плейлист ..." достаточно показывать один раз после фактического добавления
- [x] Сильно большая задержка между кликом на плейлист и скрытием bottom sheet
- [x] View типа drag handle толще, чем в макетах

https://github.com/user-attachments/assets/1ef8c880-c6d3-490f-8728-57a5c27c65cd

DmitrijEgorow commented on app/src/main/java/com/practicum/playlistmaker/di/InteractorModule.kt:
```diff
@@ -49,7 +51,13 @@ val interactorModule = module {
+        )
+    }
+
+    single<PlaylistInteractor> {
```
Интеракторы я бы делал как factory, нам не нужно для них поддерживать один инстанс

DmitrijEgorow commented on app/src/main/java/com/practicum/playlistmaker/medialibrary/presenation/newplaylist/NewPlaylistViewModel.kt:
```diff
@@ -0,0 +1,77 @@
+import com.practicum.playlistmaker.medialibrary.domain.model.Playlist
+import com.practicum.playlistmaker.search.domain.db.PlaylistInteractor
+import kotlinx.coroutines.launch
+import java.io.File
```
Работа с файлами в presentation-слое

DmitrijEgorow commented on app/src/main/java/com/practicum/playlistmaker/medialibrary/presenation/newplaylist/NewPlaylistViewModel.kt:
```diff
@@ -0,0 +1,77 @@
+package com.practicum.playlistmaker.medialibrary.presenation.newplaylist
```
typo
```suggestion
package com.practicum.playlistmaker.medialibrary.presentation.newplaylist
```

DmitrijEgorow commented on app/src/main/java/com/practicum/playlistmaker/search/data/db/dao/PlaylistDao.kt:
```diff
@@ -0,0 +1,25 @@
+@Dao
+interface PlaylistDao {
+
+    @Insert()
```
```suggestion
    @Insert
```
немного красивее будет

DmitrijEgorow commented on app/src/main/java/com/practicum/playlistmaker/utils/WordUtils.kt:
```diff
@@ -0,0 +1,14 @@
+
+object WordUtils {
+
+    fun getDeclension(count: Int, wordForms: List<String>): String {
```
Можно воспользоваться [string plurals](https://developer.android.com/guide/topics/resources/string-resource#Plurals)

pyclan88 committed:
sprint 22 playlist (исправления)

- исправил аварийное завершение приложения после выхода из Создания плейлиста через плеер
- "Добавлено в плейлист..." теперь показывается только 1 раз
- странное поведение клавиатуры в поиске у меня не воспроизводится. Очень надеюсь, что это из-за внедренной фичи, которая скрывает клавиатуру после получения результата. Удалил этот код
- исправил высоту drag handle
- переделал все интеракторы на factory
- перевел работу с файлами в data-слой
- исправил опечатку presenation
- вынес создание плейлиста в domain-слой как бизнес-логику
- сделал инсерты в PlaylistDao красивее)
- не смог внедрить string plurals, т.к. происходит неправильное спряжение, если Андроид на английском. Как разобраться с этим не знаю
